src/Heatmap/HeatMapInterpolationPoints.swift

```diff
@@ -0,0 +1,340 @@
+/* Copyright (c) 2020 Google Inc.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+import Foundation
+import GooglePlaces
+import GoogleMaps
+import GoogleMapsUtils
+
+/// This class will create artificial points in surrounding locations with appropriate intensities interpolated by neighboring intensity values.
+class HeatMapInterpolationPoints {
+
+    /// The input data set
+    private var data = [GMUWeightedLatLng]()
+
+    /// The heat map layer
+    private let heatMapLayer: GMUHeatmapTileLayer = GMUHeatmapTileLayer()
+
+    /// The heat map colors and gradient
+    private var gradientColors = [UIColor.green, UIColor.red]
+    private var gradientStartPoints = [0.005, 0.7] as [NSNumber]
+
+    /// The list of interpolated heat map points with weight
+    private var heatMapPoints = [GMUWeightedLatLng]()
+
+    /// Indicates the number of times k-means clustering should execute; will be set in the constructor to 25 by default
+    private var clusterIterations: Int!
+
+    /// Normalizing factors to convert from 2D to longitude and latitude
+    private let normalLat = 175.9783070993
```

Comment:
I wasn't sure how to convert GQTpoint to lat/long, so I found these numbers via trial/error. The longitude value makes sense to me since it ranges from -180 to 180, so multiplying makes sense (e.g. 0.5 -> 90), but the the latitude conversion is a bit weird. I found it to be consistent for multiple lat/long and GQTPoint pairs, but I would have expected the conversion to be 90 instead of 175.978. I'll note this in the code!

Reply:
Would using `GMSProjection` (which you obtain from `GMSMapView`), which projects/unprojects from CGPoint to CLLocationCoordinate2D and vice versa, work here?